src/services/KeywordExtractionService.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Extracts keywords of given PDF document. Keywords are
+ * extracted from metadata of the PDF or, if they are not
+ * found there, they are extracted from 'Keywords' section
+ * in text of PDF document. If no keywords are found, empty
+ * array is returned.
+ * @async
+ * @param {PDFDocumentProxy} pdfDoc PDF document
+ * @returns {String[]} keywords array
+ */
+async function extractKeywords(pdfDoc) {
+  let keywords = await getKeywordsFromMetadata(pdfDoc);
+
+  if (!keywords) keywords = await getKeywordsFromText(pdfDoc);
+
+  console.log(keywords); //TODO: remove
+
+  return keywords;
+}
+
+/**
+ * Extracts keywords from metadata of given PDF document.
+ * If keywords are not in the PDF document undefined returned.
+ * @async
+ * @param {PDFDocumentProxy} pdfDoc PDF document
+ * @returns {String[]} keywords array
+ */
+async function getKeywordsFromMetadata(pdfDoc) {
+  const metadata = await pdfDoc.getMetadata();
+  
+  if (metadata.info.Keywords) return metadata.info.Keywords.split(", ");
+
+  console.warn("Keywords not in metdata!");
+}
+
+/**
+ * Extracts keywords from 'Keywords' section in text of PDF document.
+ * If no keywords are found, empty array is returned.
+ * @async
+ * @param {PDFDocumentProxy} pdfDoc PDF document
+ * @returns {String[]} keywords array
+ */
+async function getKeywordsFromText(pdfDoc) {
+  const parsedKeywords = await parseKeywordsFromText(pdfDoc);
+  let keywords = new Array();
+
+  if (parsedKeywords) {
+    parsedKeywords.filter((parsed) => parsed.includes(","))
+        .forEach((parsed) => {
+            parsed.split(",")
+                .filter((keyword) => keyword.trim() != "")
+                .forEach((keyword) => keywords.push(keyword.trim()));
+        });
+
+    keywords = keywords.concat(parsedKeywords.filter((parsed) => !parsed.includes(",")));
+  }
+
+  return keywords;
+}
+
+/**
+ * Searches for text containing keywords in the text of PDF
+ * document. If there is no 'Keywords' section in text, 
+ * undefined returned, otherwise returns array of strings
+ * where each string can contain more than one keyword.
+ * @async
+ * @param {PDFDocumentProxy} pdfDoc PDF document
+ * @returns {String[]} array of strings containing keywords
+ */
+async function parseKeywordsFromText(pdfDoc) {
```

Comment:
Woah nice, did not know you would do this, does it work on some pdfs?

Reply:
If it is a "legit" research paper Keywords section is at the bottom of first page after Abstract in "Keywords: kw1, kw2, kw3, ..., kwn." format, I only have one such PDF with which I tested it and it works. Such research papers usually also have keywords in metadata so parsing them from text is not needed. It would be better if I had more such PDFs (with Keywords section) to test it with, but I also found that some other well written research papers do not have a Keywords section but have keywords in metadata, which is even better on our part.

Reply:
Great!